### PR TITLE
External references resolution

### PR DESCRIPTION
**Pull request by ildarisaev**

### CODE REVIEW

File: Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs [hunk 1]
```diff
@@ -10,124 +10,135 @@
 
 namespace Srclib.Nuget
 {
-  class DepresolveConsoleCommand
-  {
-    static Lazy<string> _dnuPath;
-
-    public static void Register(CommandLineApplication cmdApp, Microsoft.Extensions.PlatformAbstractions.IApplicationEnvironment appEnv, Microsoft.Extensions.PlatformAbstractions.IRuntimeEnvironment runtimeEnv)
+    class DepresolveConsoleCommand
```
Comment (beyang) [inline Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs:13]: Add a comment describing what this does.

File: Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs [hunk 2]
```diff
@@ -10,124 +10,135 @@
...
-          {
-            result.Add(Resolution.FromLibrary(dep));
-          }
-
-          Console.WriteLine(JsonConvert.SerializeObject(result, Formatting.Indented));
-          return 0;
-        });
-      });
-    }
+        static Lazy<string> _dnuPath;
```
Comment (beyang) [inline Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs:15]: add a comment

File: Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs [hunk 3]
```diff
@@ -10,124 +10,135 @@
...
-      var allDeps = GetAllDeps(proj);
-      if (allDeps.Any(dep => !dep.Resolved))
-      {
-        await RunResolve(dir);
-        allDeps = GetAllDeps(proj);
-      }
-
-      return allDeps;
-    }
+        public static void Register(CommandLineApplication cmdApp, Microsoft.Extensions.PlatformAbstractions.IApplicationEnvironment appEnv, Microsoft.Extensions.PlatformAbstractions.IRuntimeEnvironment runtimeEnv)
```
Comment (beyang) [inline Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs:23]: Add a comment describing functionality

File: Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs [hunk 4]
```diff
@@ -10,124 +10,135 @@
...
+
+            p.Start();
+            var result = p.StandardOutput.ReadToEnd().Trim();
+            p.WaitForExit();
+            return result;
+        }
+
+        public static string GetDll(string path)
+        {
+            return RunForResult("/bin/bash", "-c \"cd " + path + " && find -name *.dll | head -1\"");
```
Comment (beyang) [inline Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs]: Can this be done in code? A little hacky to shell out to a bash utility here.

File: Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs [hunk 5]
```diff
@@ -10,124 +10,135 @@
...
+            p.Start();
+            // it's important to read stdout and stderr, else it might deadlock
+            var outs = await Task.WhenAll(p.StandardOutput.ReadToEndAsync(), p.StandardError.ReadToEndAsync());
+            p.WaitForExit();
+            // in the future, actually parse output or something
+        }
+
+        static string FindDnuWindows()
+        {
+            return RunForResult(@"C:\Windows\System32\cmd.exe", "/c \"where dnu\"");
```
Comment (beyang) [inline Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs]: how robust is this on Windows? Will it always work?

File: Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs [hunk 6]
```diff
@@ -10,124 +10,135 @@
...
+        }
 
-    static string FindDnuWindows()
-    {
-      return RunForResult(@"C:\Windows\System32\cmd.exe", "/c \"where dnu\"");
-    }
+        static string FindDnuNix()
+        {
+            return RunForResult("/bin/bash", "-c \"which dnu\"");
```
Comment (beyang) [inline Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs:115]: There must be some way to look for this in the system path in C#.

File: Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs (hunk 2 above)
Comment (beyang) [inline Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs:15]: If we assume dnu is on the system path, anyway (which we do, because the way we compute this value is by doing `which dnu` or `where dnu`), then why is it even necessary to get the absolute path? Can't we just call `dnu` and it will work because it's on the system path?

File: Srclib.Nuget/Graph/GraphRunner.cs [hunk 7]
```diff
@@ -14,185 +14,297 @@
...
+        private GraphRunner() : base(SyntaxWalkerDepth.Token)
+        {
+            projectUrlToRepo["http://www.newtonsoft.com/json"] = "github.com/JamesNK/Newtonsoft.Json";
+            projectUrlToRepo["http://autofac.org/"] = "github.com/autofac/Autofac";
+            projectUrlToRepo["http://msdn.com/roslyn"] = "github.com/dotnet/roslyn";
+            projectUrlToRepo["http://www.asp.net/"] = "github.com/aspnet/dnx";//this may not be always correct
+        }
 
-        if (_defined.Contains(definition))
+        /// <summary>
```
Comment (beyang) [inline Srclib.Nuget/Graph/GraphRunner.cs:49]: Let's follow the MSDN C# style guidelines when it comes to comments and everything else in the code: https://msdn.microsoft.com/en-us/library/ff926074.aspx

File: Srclib.Nuget/Graph/GraphRunner.cs [hunk 8]
```diff
@@ -14,185 +14,297 @@
...
+                    {
+                        var reference = Ref.To(definition).At(file, token.Span);
+                        _output.Refs.Add(reference);
+                    }
+                    else
+                    {
+                        _defined.Add(definition);
+                        //at first encountering
+                        var def = Def.For(symbol: definition, type: "external", name: definition.Name).At(file, token.Span);
+                        //_output.Defs.Add(def);
```
Comment (beyang) [inline Srclib.Nuget/Graph/GraphRunner.cs]: delete

File: Srclib.Nuget/Graph/GraphRunner.cs [hunk 9]
```diff
@@ -14,185 +14,297 @@
...
+                    var file = r.Item3;
+                    if (_defined.Contains(definition))
+                    {
+                        var reference = Ref.To(definition).At(file, token.Span);
+                        _output.Refs.Add(reference);
+                    }
+                    else
+                    {
+                        _defined.Add(definition);
+                        //at first encountering
```
Comment (beyang) [inline Srclib.Nuget/Graph/GraphRunner.cs]: either make this is a readable sentence or delete

File: Srclib.Nuget/Graph/GraphRunner.cs [hunk 10]
```diff
@@ -14,185 +14,297 @@
...
+                        if ((definition.ContainingAssembly != null) && (definition.ContainingAssembly.Identity != null) && (definition.ContainingAssembly.Identity.Name != null))
+                        {
+                            if (dllToProjectUrl.ContainsKey(definition.ContainingAssembly.Identity.Name + ".dll"))
+                            {
+                                string url = dllToProjectUrl[definition.ContainingAssembly.Identity.Name + ".dll"];
+                                if (projectUrlToRepo.ContainsKey(url))
+                                {
+                                    url = projectUrlToRepo[url];
+                                }
+                                //filter out non cloneable urls
```
Comment (beyang) [inline Srclib.Nuget/Graph/GraphRunner.cs]: comment syntax

Comment (beyang) [inline Srclib.Nuget/Graph/GraphRunner.cs]: can we have some kind of C# linter to enforce style guidelines?

File: Srclib.Nuget/Graph/GraphRunner.cs [hunk 11]
```diff
@@ -14,185 +14,297 @@
...
-      }
 
-      runner._sm = null;
-      runner._path = null;
-      runner.RunTokens();
-      return runner._output;
-    }
+        internal static void HandleNuspec(string path, string dll)
+        {
+            //for some reason xml parsing libraries doesn't seem to be available for dnx50
```
Comment (beyang) [inline Srclib.Nuget/Graph/GraphRunner.cs]: Rephrase this to be more descriptive and less casual. E.g., "Parse nuspec manually, because XML parsing libraries aren't available for dnx50".

What is dnx50? Explain that, as well.

File: Srclib.Nuget/Graph/GraphRunner.cs [hunk 12]
```diff
@@ -14,185 +14,297 @@
...
-        if (!_defined.Contains(symbol))
-        {
-          _defined.Add(symbol);
+            context.CompilationEngine = new CompilationEngine(context.CompilationContext);
+            context.LibraryExporter = context.CompilationEngine.CreateProjectExporter(context.Project, context.ApplicationHostContext.TargetFramework, "Debug");
+            context.Export = context.LibraryExporter.GetExport(context.Project.Name);
 
-          var def = Def.For(symbol: symbol, type: "interface", name: symbol.Name)
-            .At(_path, node.Identifier.Span);
+            //If other languages are to be supported, this part needs to change.
```
Comment (beyang) [inline Srclib.Nuget/Graph/GraphRunner.cs]: comment syntax

**Comment by beyang** [comment]
Please add a high-level summary of the changes and what to look for. Otherwise, I cannot do an adequate review.

Please explain the changes to the test files, too. Want to make sure no major regressions were introduced there. Given that it looks like there were big changes, please explain why they are correct.

File: Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs (hunk 5 above)
Comment (ildarisaev) [inline Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs]: All this dnu activity came from code written by @Alxandr. I kept it intact. I think we can remove anything related to Windows here.

**Comment by ildarisaev** [comment]
Most of the meaningful changes are located in GraphRunner.cs. I added the code that runs the resolution of project dependencies, loads the corresponding dll files and adds them to the roslyn compilation (so it can read semantic info out of these dlls). This code is located in Graph method between lines 183 and 235.

I added the code that reads projectUrl attribute out of nuspec file. This is located in method HandleNuspec.

I also added the code that handles references without definitions (e. g. the external references). It generates a ref and tries to set up a link to another project. This part of the changes is located in RunTokens method, between lines 89 and 117.

I added a map from project urls to the urls of corresponding repositories. It is initialized at GraphRunner constructor at lines 38-41. I think it will contain more records in the future.

I added some code in VisitToken method between lines 551-565. Sometimes roslyn compiler doesn't resolve a token to a single semantic entity but to a group of entities instead. Previously this situation wasn't handled at all, the token was just skipped, no refs was generated. Now i set up a ref to the top entity in the group.

As for the tests. Tests require much improvement. With initial @Alxandr 's prototype, just some not very big c# project was used as a testdata. At first I didn't changed that, just pushed it into sgtest. At some moment while adding external references resolution functionality i observed different test behaviour at my local machine and in travis environment. Like, srclib test produced different output on my machine and in travis, and thus failed. Now i think that it may have been caused by this concurrency issue that i wrote you about, but i didn't know that at those moment, i understood it later. Travis environment is impossible to debug, and real life project wasn't a best choice for testing anyway, cause test results are hard to observe and evaluate. So at some point i threw away old testdata and added some simple stub testdata instead. I plan to add some testdata that would result in something closer to unit tests instead of regression tests. Like, there should be different language constructions to check, self-references, external references, probably some tests projects with references to each other etc.

**Comment by beyang** [comment]
Test cases to check specific language features / srclib functionality (like external references) would be fantastic. Can we add some of those after this change is merged?

File: Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs (hunk 4 above)
Comment (ildarisaev) [inline Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs]: Yes, it is hacky indeed. For some reason, dnx interface doesn't resolve a dependency to an actual dll file. Instead it resolves a dependency to a some folder, that contains multiple same-named dlls for different platforms (and sometimes different versions), e. g.

~/.dnx/packages/System.Reflection/4.0.10# find -name "*.dll"

./lib/DNXCore50/System.Reflection.dll
./lib/netcore50/System.Reflection.dll
./ref/dotnet/System.Reflection.dll
./runtimes/win8-aot/lib/netcore50/System.Reflection.dll

This is really inconvenient, but I didn't find any straightforward ways to deal with it in dnx interface. I'll try to dig in roslyn sources to see how it solves this problem (obviously roslyn needs to solve exactly the same problem). But for now i have to do it through bash invocation.

**Comment by ildarisaev** [comment]
Yes, I hope to do it, meaningful tests would be very useful.

File: Srclib.Nuget/Graph/GraphRunner.cs (hunk 11 above)
Comment (ildarisaev) [inline Srclib.Nuget/Graph/GraphRunner.cs]: It should be dnxcore50, not dnx50. With introduction of dnx, c# applications can be run on different operating systems with different runtimes. That made things much easier, but unfortunately different runtimes have different libraries with different interfaces, and it makes it harder to write universal code that can be compiled everywhere.

It is also somehow explained here http://stackoverflow.com/questions/31654505/dnx-vs-dnx-core

To say it simple, i didn't find an xml parsing library for runtime we use to run toolchain in linux environment. Anyway, the comment was quite meaningless there, i just threw it away.

File: Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs (hunk 4 above)
Comment (beyang) [inline Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs]: Right, but what I mean was, instead of using `find`, could you just walk the file path in code and return the first file with a `.dll` extension? That way, you don't have any Unix-specific code and you don't have to rely on `/bin/bash` existing.

File: Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs [hunk 13]
```diff
@@ -10,124 +10,129 @@
...
-      p.StartInfo.RedirectStandardOutput = true;
-      p.StartInfo.RedirectStandardError = false;
-
-      p.Start();
-      var result = p.StandardOutput.ReadToEnd().Trim();
-      p.WaitForExit();
-      return result;
+        public static string GetDll(string path, string name)
+        {
+            return RunForResult("/bin/bash", "-c \"cd " + path + " && find -name " + name + ".dll | head -1\"");
```
Comment (beyang) [inline Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs]: same here, instead of shelling out, you could do this in code. See https://msdn.microsoft.com/en-us/library/bb513869.aspx

**Comment by beyang** [comment]
Apart from the shelling out to `/bin/bash -c find` and the comment syntax question, it looks good to me.

File: Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs (hunk 4 above)
Comment (ildarisaev) [inline Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs]: i see your point now!)
